pallets/parachain-staking/src/nominator_list.rs

```diff
@@ -0,0 +1,198 @@
+// Copyright 2019-2021 PureStake Inc.
+// This file is part of Moonbeam.
+
+// Moonbeam is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Moonbeam is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Moonbeam.  If not, see <http://www.gnu.org/licenses/>.
+
+#![cfg_attr(not(feature = "std"), no_std)]
+
+use crate::*;
+use sp_runtime::{
+	traits::AtLeast32BitUnsigned,
+};
+use sp_std::prelude::*;
+
+/// An in-memory struct representing a collator's list of nominators. This list is sorted by amount
+/// of stake upon every iteration and the top N nominators are trivially available at any time. In
+/// addition, several sums are maintained (also updated upon any mutation):
+///
+/// 1) the total sum of all nominators
+/// 2) the sum of the top N nominators
+/// 3) the sum of the nominators outside of the top N (the "bottom" nominators)
+///
+/// Design note: I find it to be a useful exercise to fully abstract away specifics. This first pass
+/// doesn't reflect this, but consider what this struct and its functions might look like if they
+/// had nothing to do with collators or nominators.
+pub struct CollatorNominators<AccountId, Balance>  {
+	/// Sorted list of all nominators for this collator
+	pub nominators: Vec<AccountId>,
+	/// All bonds, sorted by bonded amount
+	pub bonds: Vec<Bond<AccountId, Balance>>,
+	/// Maximum number of nominators that will be selected
+	pub max_selected: u32,
```

Comment:
This is a global constant, I do not think we want this configurable per collator

Reply:
One way to convey that is `pub struct CollatorNominators<AccountId, Balance, MaxSelected: Get<u32>>`

Reply:
Good idea, Joshy. It's not meant to be configurable per-collator, so it's at least confusing to see it exist as a struct member.

Reply:
Then you can plumb it through from the configuration trait with something like `type CollatorNominatorsOf<T: Config> = CollatorNominators<T::AccountId, BalanceOf<T>, T::MaxSelected>;`.

Also this struct needs a better name. What about `DelegationInfo` (starting to switch toward dpos language).

Reply:
Is the nominator data duplicated in this struct?

Reply:
Yes, nominator data is duplicated. This is because we want to be able to find an item (by `AccountId`) efficiently but also be able to iterate through `Bond`s in order of bond amount.

Your observation about this resembling a PriorityQueue is intriguing. I think we'll end up with something similar (where we have at least two overlapping data structures) if we extend our requirements to a PQ:

* efficiently detect that an item exists (item == AccountId)
* efficiently iterate through items in order of bond amount
* efficiently shift items in and out of some `top-N sum` (?)

Our current implementation (both in master and in this PR as of me writing this) is to use basically a heap, and I think this is where so much of the staking complication comes from; we implement a heap in this crate manually on top of a vec (plus all of the sorting at every op on it).

I had considered trying to use a single, sorted vec and a bloom filter for our first requirement, but a BF seems like the sort of thing that wouldn't hold up in a paranoid/secure situation.

This PQ: https://docs.rs/priority-queue/ uses a hash table, which could be well suited for our efficient lookup by `AccountId` requirement. It's certainly worth exploring.

However, consider that the most frequent pattern of use is something like:

1. query data from DB
2. deserialize through SCALE
3. operate (add/remove nomination, shift nomination's "priority")
4. serialize back through SCALE
5. flush to our DB

Given this pattern, a heap might be hard to beat. But I'm also in favor of leaving some optimizations for later...

Reply:
Sorry, I was conflating a couple things in my response. Allow me to clarify:

* the existing code manually sorts the lists of nominators (nothing to do with a heap)
* the lists are stored in sorted order on chain
    * however, it is not a stable-sort, so it may rearrange equal items
* an improvement is partially-introduced in this PR: items are inserted in-order, so no sorting should ever be necessary
    * so an insert is a binary search followed by a shift of elements
    * (TODO: make sure an insert when equal items exist inserts at the end)

I believe this is optimal for dealing with bond amounts. Every operation we perform on this data structure can be done in a way that maintains sorted order. Removing or rearranging items isn't very efficient, however, as this requires a shift of up to N elements.

A heap would incur at least one major disadvantage: it's not efficient for ordered iteration. 

After some consideration, I think we should remove the sorted `nominators` Vec. Its purpose is to efficiently detect if a nominator exists within a collator -- basically, to avoid a scan. We have a few choices for this:

1. Maintain an (e.g. sorted) data structure for this and store on-chain
2. Build a fresh data structure each time we need it
3. Scan

Of these, I think scan is probably the most efficient, sadly. Alternatively, we may be able to use `NominatorState` for a more efficient way to look this up.

With this simplification, this data structure basically becomes (1) a sorted list and (2) some cached values.

Reply:
> After some consideration, I think we should remove the sorted nominators Vec. Its purpose is to efficiently detect if a nominator exists within a collator -- basically, to avoid a scan. We have a few choices for this:

This is a space optimization right? For any optimization, it'd be nice to see the benchmarking results diff.

Reply:
Agreed. Basically, my conclusion is that we either need to store a sorted list on chain or do an inefficient scan, and I suspect that the latter is actually more efficient. Definitely worth measuring, though.

One thing that could help is to store a much smaller bloom filter on chain, but that probably doesn't help because of the "we charge for the worst case anyway" problem.

Reply:
> Basically, my conclusion is that we either need to store a sorted list on chain or do an inefficient scan, and I suspect that the latter is actually more efficient.

It is more memory efficient, but less computationally efficient to search for membership the larger the vector grows. This is the idea behind using a single, ordered `Vec` for the `.nominators` set.

It would be easy to make the aforementioned change to `Collator2` struct and then run benchmarks to see if anything changes. I'd just remove the `.nominators` field and replace all the places where we search it with code that goes over the entire set of nominations to see if the nomination exists.

Reply:
Good point. We could probably copy an existing fn but have it do a scan (totally ignore the on-chain sorted list) and then benchmark it and its original to see how they compare.